File: RRT_Test/python/rrtstar.py
import numpy as np
import cv2
import heapq
import time
from math import dist

WHEEL_RADIUS = 33#/1000 # 33mm
ROBOT_RADIUS = 220#/1000 # 220mm
WHEEL_DISTANCE = 287#/1000 # 287mm
clearance = 5
clearance += ROBOT_RADIUS

width = 6000
height = 2000
scale = 5

clearance_color = (0, 255, 255)
obstacle_color = (0, 0, 0)
robot_radius_color = (254, 105, 180)

# Create a black canvas
canvas = np.zeros((height, width, 3), dtype="uint8")
# Create a white rectangle
canvas = cv2.rectangle(canvas, (clearance, clearance), (width-clearance, height-clearance), (255, 255, 255), -1)

# OBSTACLE 1
x1, x2 = 1500, 1750
y1, y2 = 0, 1000
# Draw the clearance
for i in range(x1-clearance, x2+clearance):
    for j in range(y1+clearance, y2+clearance):
        canvas[j, i] = clearance_color
# Draw the obstacle
for i in range(x1, x2):
    for j in range(y1, y2):
        canvas[j, i] = obstacle_color

# OBSTACLE 2
x1, x2 = 2500, 2750
y1, y2 = height-1000, height
# Draw the clearance
for i in range(x1-clearance, x2+clearance):
    for j in range(y1-clearance, y2-clearance+1):
        canvas[j, i] = clearance_color
# Draw the obstacle
for i in range(x1, x2):
    for j in range(y1, y2):
        canvas[j, i] = obstacle_color

# OBSTACLE 3
# Draw a circle at (4200, 800)
center = (4200, 800)
radius = 600
# Draw the clearance
canvas = cv2.circle(canvas, center, 600+clearance, clearance_color, -1)
# Draw the obstacle
canvas = cv2.circle(canvas, center, 600, obstacle_color, -1)
        
# x_start, y_start = clearance+1, clearance+1
# x_goal, y_goal = width-clearance-1, clearance+1

# Draw a circle at x_start, y_start
# canvas = cv2.circle(canvas, (x_start, y_start), 5, (0, 255, 0), -1)
# Draw a circle at x_goal, y_goal
# canvas = cv2.circle(canvas, (x_goal, y_goal), 5, (0, 0, 255), -1)


# Draw a red line at x = width/2
# canvas = cv2.line(canvas, (int(width/2.5), 0), (int(width/2.5), height), (0, 0, 255), 10)
# Resize the canvas by a factor of scale
width_resized = int(width/scale)
height_resized = int(height/scale)
canvas_resized = cv2.resize(canvas, (width_resized, height_resized))
# cv2.imshow("Canvas", canvas_resized)
# cv2.waitKey(0)
# cv2.destroyAllWindows()

def obstacles():
    width = 6000
    height = 2000
    scale = 5

    clearance_color = (0, 255, 255)
    obstacle_color = (0, 0, 0)

    # Create a black canvas
    canvas = np.zeros((height, width, 3), dtype="uint8")
    # Create a white rectangle
    canvas = cv2.rectangle(canvas, (clearance, clearance), (width-clearance, height-clearance), (255, 255, 255), -1)

    # OBSTACLE 1
    x1, x2 = 1500, 1750
    y1, y2 = 0, 1000
    # Draw the clearance
    for i in range(x1-clearance, x2+clearance):
        for j in range(y1+clearance, y2+clearance):
            canvas[j, i] = clearance_color
    # Draw the obstacle
    for i in range(x1, x2):
        for j in range(y1, y2):
            canvas[j, i] = obstacle_color

    # OBSTACLE 2
    x1, x2 = 2500, 2750
    y1, y2 = height-1000, height
    # Draw the clearance
    for i in range(x1-clearance, x2+clearance):
        for j in range(y1-clearance, y2-clearance+1):
            canvas[j, i] = clearance_color
    # Draw the obstacle
    for i in range(x1, x2):
        for j in range(y1, y2):
            canvas[j, i] = obstacle_color

    # OBSTACLE 3
    # Draw a circle at (4200, 800)
    center = (4200, 800)
    radius = 600
    # Draw the clearance
    canvas = cv2.circle(canvas, center, 600+clearance, clearance_color, -1)
    # Draw the obstacle
    canvas = cv2.circle(canvas, center, 600, obstacle_color, -1)
            
    # x_center, y_center = 4200, 800
    # radius = 600
    # Draw the clearance
    # for i in range(x_center-radius-clearance, x_center+radius+clearance):
    #     for j in range(y_center-radius-clearance, y_center+radius+clearance):
    #         if (i-x_center)**2 + (j-y_center)**2 <= (radius+clearance)**2 and canvas[j, i, 0] != 0:
    #             canvas[j, i] = clearance_color
    # # Draw the obstacle
    # for i in range(x_center-radius, x_center+radius):
    #     for j in range(y_center-radius, y_center+radius):
    #         if (i-x_center)**2 + (j-y_center)**2 <= radius**2:
    #             canvas[j, i] = obstacle_color

    # x_start, y_start = clearance+1, clearance+1
    # x_goal, y_goal = width-clearance-1, clearance+1

    # Draw a circle at x_start, y_start
    # canvas = cv2.circle(canvas, (x_start, y_start), 5, (0, 255, 0), -1)
    # Draw a circle at x_goal, y_goal
    # canvas = cv2.circle(canvas, (x_goal, y_goal), 5, (0, 0, 255), -1)


    # Draw a red line at x = width/2
    # canvas = cv2.line(canvas, (int(width/2.5), 0), (int(width/2.5), height), (0, 0, 255), 10)
    # Resize the canvas by a factor of scale
    width_resized = int(width/scale)
    height_resized = int(height/scale)
    canvas_resized = cv2.resize(canvas, (width_resized, height_resized))
    # cv2.imshow("Canvas", canvas_resized)
    # cv2.waitKey(0)
    # cv2.destroyAllWindows()
    return canvas


canvas = obstacles()

x_start, y_start, theta_start = clearance+1, clearance+1, 0
x_goal, y_goal = width-clearance-1, clearance+1

step_size = 100
distance_threshold = 100
search_radius = 300

# Make a lambda function to adjust the value of x to the visited space
adjust = lambda x, threshold: int(int(round(x*2)/2)/threshold)

# Dictionary to store visited nodes
visited = {(adjust(x_start, distance_threshold),
            adjust(y_start, distance_threshold)): 1}

# Dictionary to store the parent of each node
parent = {(x_start, y_start): (x_start, y_start)}

# Dictionary to store the children of each node
children = {(x_start, y_start): []}

# Dictionary to store the cost of each node
cost = {(x_start, y_start): 0}

# Function to check if the node is valid
def valid_node(x, y):
    # Adjust the value for visited nodes
    x_vis = adjust(x, distance_threshold)
    y_vis = adjust(y, distance_threshold)
    # Adjust the values for canvas
    x_cvs = int(round(x*2)/2)
    y_cvs = int(round(y*2)/2)

    # Check if the node is not in the obstacle and not visited
    if canvas[y_cvs, x_cvs, 0] != 0 and (x_vis, y_vis) not in visited:
        return True
    return False

# Function to check if the child node and parent node are not intersecting with the obstacle
def valid_edge(x_parent, y_parent, x_child, y_child):
    # Sample 3 points on the line joining the parent and child nodes, 
    # not including the parent and child nodes
    x_intermediate = np.linspace(x_parent, x_child, 5)[1:-1]
    y_intermediate = np.linspace(y_parent, y_child, 5)[1:-1]

    # Adjust the values for canvas
    x_intermediate = [int(round(x*2)/2) for x in x_intermediate]
    y_intermediate = [int(round(y*2)/2) for y in y_intermediate]

    # Check if any of the intermediate points are in the obstacle
    for x, y in zip(x_intermediate, y_intermediate):
        if canvas[y, x, 0] == 0:
            return False
    return True


reached = False
iterations = 0
c_best = float('inf')
start = time.time()

while iterations < 100:
    
    # Get a new node
    x_node = np.random.randint(clearance, width-clearance)
    y_node = np.random.randint(clearance, height-clearance)    

    # If the node is valid
    if valid_node(x_node, y_node):

        # Find the nearsest node in the tree to get the projected node
        min_dist = float('inf')

        for x, y in parent:
            dist = np.sqrt((x-x_node)**2 + (y-y_node)**2)
            if dist < min_dist:
                min_dist = dist
                x_parent, y_parent = x, y

        # Get the angle of the new node
        theta = np.arctan2(y_node-y_parent, x_node-x_parent)

        # Calculate the new node
        x_new = int(x_parent + step_size*np.cos(theta))
        y_new = int(y_parent + step_size*np.sin(theta))
        
        if reached:
            iterations += 1

        # Check if the new node is valid
        if valid_node(x_new, y_new):

            # Collect all the nodes in the search radius
            neighbours = []
            for x, y in parent:
                if np.sqrt((x-x_new)**2 + (y-y_new)**2) < search_radius:
                    neighbours.append((x, y))

            # Find the node with the minimum cost to get new node
            min_cost = cost[(x_parent, y_parent)] + step_size
            for x, y in neighbours:
                new_cost = cost[(x, y)] + np.sqrt((x-x_new)**2 + (y-y_new)**2)
                if new_cost < min_cost:
                    min_cost = new_cost
                    x_parent, y_parent = x, y

            # Check if the edge between the parent and child nodes is valid
            if not valid_edge(x_parent, y_parent, x_new, y_new):
                continue
            
            # Add the cost of the new node
            cost[(x_new, y_new)] = min_cost

            # Check if rewiring to the newly added node will reduce the cost of the neighbours
            for x, y in neighbours:
                new_cost = cost[(x_new, y_new)] + np.sqrt((x-x_new)**2 + (y-y_new)**2)
                if new_cost < cost[(x, y)]:

                    # Check if the edge between the parent and child nodes is valid
                    if not valid_edge(x, y, x_new, y_new):
                        continue

                    previous_cost = cost[(x, y)]
                    cost[(x, y)] = new_cost
                    
                    # Remove this node from the children of its current parent
                    x_parent_neighbour, y_parent_neighbour = parent[(x, y)]
                    children[(x_parent_neighbour, y_parent_neighbour)].remove((x, y))

                    # Add the new node to the parent dictionary
                    parent[(x, y)] = (x_new, y_new)

                    # Add the new node to the children dictionary
                    children.setdefault((x_new, y_new), []).append((x, y))
                    
                    # If the cost of this node is reduced, we must update the cost of its children,
                    # and their children, and so on
                    cost_reduction = previous_cost - new_cost
                    queue = []
                    heapq.heappush(queue, ((x, y)))
                    while queue:
                        x_, y_ = heapq.heappop(queue)
                        # Check if the node has children
                        if (x_, y_) in children:
                            for child in children[(x_, y_)]:
                                    cost[child] -= cost_reduction
                                    heapq.heappush(queue, child)
                                    
            x_achieved, y_achieved = x_new, y_new

            x_adjusted = adjust(x_new, distance_threshold)
            y_adjusted = adjust(y_new, distance_threshold)
            
            # Add the new node to the visited nodes
            visited[(x_adjusted, y_adjusted)] = 1

            # Add the new node to the parent dictionary
            parent[(x_new, y_new)] = (x_parent, y_parent)

            # Add the new node to the children dictionary
            children.setdefault((x_parent, y_parent), []).append((x_new, y_new))

            if not reached:
                # Check if the new node is close to the goal
                if np.sqrt((x_new-x_goal)**2 + (y_new-y_goal)**2) < distance_threshold:
                # if np.sqrt((x_new-x_goal)**2 + (y_new-y_goal)**2) < 50:
                    end = time.time()
                    print("Goal reached: ", end-start, "seconds")

                    # Add the final node as a child of the current node to complete the path
                    children.setdefault((x_new, y_new), []).append((x_goal, y_goal))

                    # Add the final node to the parent dictionary
                    parent[(x_goal, y_goal)] = (x_new, y_new)

                    # Save the final node
                    x_final, y_final = x_goal, y_goal

                    # Add the cost of the final node
                    cost[(x_goal, y_goal)] = cost[(x_new, y_new)] + np.sqrt((x_goal-x_new)**2 + (y_goal-y_new)**2)

                    print("Current Cost: ", cost[(x_goal, y_goal)])

                    # Set the reached flag to True
                    reached = True
        
print("Final cost: ", cost[(x_final, y_final)]) 

# Get the path from the parent dictionary
path = []
# x, y = x_goal, y_goal   
# x, y = x_achieved, y_achieved
x, y = x_final, y_final
while (x, y) != (x_start, y_start):
    # print(x, y)
    path.append((x, y))
    x, y = parent[(x, y)]
path.append((x, y))
path.reverse()

# Calculate the distance between every point in the path
sum = 0
for i in range(len(path)-1):
    # print(path[i])
    sum += np.sqrt((path[i][0]-path[i+1][0])**2 + (path[i][1]-path[i+1][1])**2)

print("Total distance: ", sum)
print("Difference in cost: ", np.round(cost[(x_final, y_final)] - sum, 2))

canvas = obstacles()

# Plot the start and goal nodes
canvas = cv2.circle(canvas, (x_start, y_start), 5, (0, 0, 254), 10) 
canvas = cv2.circle(canvas, (x_goal, y_goal), 5, (0, 0, 254), 10)

# Plot a line between the parent and child nodes
for x, y in parent:
    x_parent, y_parent = parent[(x, y)]
    canvas = cv2.line(canvas, (x, y), (x_parent, y_parent), (0, 255, 0), 5)
    # Draw a purple circle at the parent node
    canvas = cv2.circle(canvas, (x_parent, y_parent), 5, (255, 0, 255), 3)

# Plot a circle at all the child points 
for x, y in children:
    for child in children[(x, y)]:
        canvas = cv2.circle(canvas, child, 5, (255, 0, 255), 3)

# Plot the path
for i in range(len(path)-1):
    x1, y1 = path[i]
    x2, y2 = path[i+1]
    canvas = cv2.line(canvas, (x1, y1), (x2, y2), (0, 0, 255), 5)

# Resize the canvas by a factor of scale
canvas_resized = cv2.resize(canvas, (width_resized, height_resized))

cv2.imshow("Canvas", canvas_resized)
cv2.waitKey(0)
cv2.destroyAllWindows()

File: RRT_Test/src/rrt_star.cpp
#include <opencv2/opencv.hpp>
#include <iostream>
#include <vector>
#include <cmath>
#include <unordered_map>
#include <queue>
#include <algorithm>
#include <random>
#include <limits>

// 为 std::pair<int, int> 提供自定义哈希函数
struct pair_hash {
    template <class T1, class T2>
    std::size_t operator () (const std::pair<T1,T2> &p) const {
        auto hash1 = std::hash<T1>{}(p.first);
        auto hash2 = std::hash<T2>{}(p.second);
        return hash1 ^ (hash2 + 0x9e3779b9 + (hash1 << 6) + (hash1 >> 2));
    }
};

// 定义机器人和环境相关参数
const int WHEEL_RADIUS = 33;
const int ROBOT_RADIUS = 220;
const int WHEEL_DISTANCE = 287;
const int clearance = 5 + ROBOT_RADIUS;
const int width = 6000;
const int height = 2000;
const int scale = 5;

const cv::Scalar clearance_color(0, 255, 255);
const cv::Scalar obstacle_color(0, 0, 0);
const cv::Scalar robot_radius_color(254, 105, 180);

// 创建包含障碍物的画布
cv::Mat obstacles() {
    cv::Mat canvas = cv::Mat::zeros(cv::Size(width, height), CV_8UC3);
    cv::rectangle(canvas, cv::Point(clearance, clearance), cv::Point(width - clearance, height - clearance), cv::Scalar(255, 255, 255), -1);

    // 障碍物 1
    int x1 = 1500, x2 = 1750;
    int y1 = 0, y2 = 1000;
    for (int i = x1 - clearance; i < x2 + clearance; ++i) {
        for (int j = y1 + clearance; j < y2 + clearance; ++j) {
            canvas.at<cv::Vec3b>(j, i) = cv::Vec3b(clearance_color[0], clearance_color[1], clearance_color[2]);
        }
    }
    for (int i = x1; i < x2; ++i) {
        for (int j = y1; j < y2; ++j) {
            canvas.at<cv::Vec3b>(j, i) = cv::Vec3b(obstacle_color[0], obstacle_color[1], obstacle_color[2]);
        }
    }

    // 障碍物 2
    x1 = 2500; x2 = 2750;
    y1 = height - 1000; y2 = height;
    for (int i = x1 - clearance; i < x2 + clearance; ++i) {
        for (int j = y1 - clearance; j < y2 - clearance + 1; ++j) {
            canvas.at<cv::Vec3b>(j, i) = cv::Vec3b(clearance_color[0], clearance_color[1], clearance_color[2]);
        }
    }
    for (int i = x1; i < x2; ++i) {
        for (int j = y1; j < y2; ++j) {
            canvas.at<cv::Vec3b>(j, i) = cv::Vec3b(obstacle_color[0], obstacle_color[1], obstacle_color[2]);
        }
    }

    // 障碍物 3
    cv::Point center(4200, 800);
    int radius = 600;
    cv::circle(canvas, center, 600 + clearance, clearance_color, -1);
    cv::circle(canvas, center, 600, obstacle_color, -1);

    return canvas;
}

// 调整坐标以用于访问已访问节点
int adjust(int x, int threshold) {
    return static_cast<int>(std::round(x * 2) / 2) / threshold;
}

// 检查节点是否有效
bool valid_node(const cv::Mat& canvas, const std::unordered_map<std::pair<int, int>, int, pair_hash>& visited, int x, int y, int distance_threshold) {
    int x_vis = adjust(x, distance_threshold);
    int y_vis = adjust(y, distance_threshold);
    int x_cvs = static_cast<int>(std::round(x * 2) / 2);
    int y_cvs = static_cast<int>(std::round(y * 2) / 2);

    return canvas.at<cv::Vec3b>(y_cvs, x_cvs)[0] != 0 && visited.find({x_vis, y_vis}) == visited.end();
}

// 检查边是否有效
bool valid_edge(const cv::Mat& canvas, int x_parent, int y_parent, int x_child, int y_child) {
    // 采样 3 个中间点
    std::vector<int> x_intermediate;
    std::vector<int> y_intermediate;
    for (int i = 1; i < 4; ++i) {
        double t = static_cast<double>(i) / 4;
        int x = static_cast<int>(std::round((1 - t) * x_parent + t * x_child));
        int y = static_cast<int>(std::round((1 - t) * y_parent + t * y_child));
        x_intermediate.push_back(static_cast<int>(std::round(x * 2) / 2));
        y_intermediate.push_back(static_cast<int>(std::round(y * 2) / 2));
    }

    for (size_t i = 0; i < x_intermediate.size(); ++i) {
        int x = x_intermediate[i];
        int y = y_intermediate[i];
        if (canvas.at<cv::Vec3b>(y, x)[0] == 0) {
            return false;
        }
    }
    return true;
}

// 计算两点之间的距离
double distance(int x1, int y1, int x2, int y2) {
    return std::sqrt(std::pow(x1 - x2, 2) + std::pow(y1 - y2, 2));
}

int main() {
    cv::Mat canvas = obstacles();

    int x_start = clearance + 1, y_start = clearance + 1, theta_start = 0;
    int x_goal = width - clearance - 1, y_goal = clearance + 1;

    int step_size = 100;
    int distance_threshold = 100;
    int search_radius = 300;

    // 使用自定义哈希函数
    std::unordered_map<std::pair<int, int>, int, pair_hash> visited;
    visited[{adjust(x_start, distance_threshold), adjust(y_start, distance_threshold)}] = 1;

    // 使用自定义哈希函数
    std::unordered_map<std::pair<int, int>, std::pair<int, int>, pair_hash> parent;
    parent[{x_start, y_start}] = {x_start, y_start};

    // 使用自定义哈希函数
    std::unordered_map<std::pair<int, int>, std::vector<std::pair<int, int>>, pair_hash> children;
    children[{x_start, y_start}] = {};

    // 使用自定义哈希函数
    std::unordered_map<std::pair<int, int>, double, pair_hash> cost;
    cost[{x_start, y_start}] = 0;

    bool reached = false;
    int iterations = 0;
    double c_best = std::numeric_limits<double>::max();

    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_int_distribution<> dis_x(clearance, width - clearance);
    std::uniform_int_distribution<> dis_y(clearance, height - clearance);

    auto start = std::chrono::high_resolution_clock::now();
    while (iterations < 500) {
        int x_node = dis_x(gen);
        int y_node = dis_y(gen);

        if (valid_node(canvas, visited, x_node, y_node, distance_threshold)) {
            double min_dist = std::numeric_limits<double>::max();
            std::pair<int, int> nearest;
            for (const auto& p : parent) {
                double dist = distance(p.first.first, p.first.second, x_node, y_node);
                if (dist < min_dist) {
                    min_dist = dist;
                    nearest = p.first;
                }
            }
            int x_parent = nearest.first;
            int y_parent = nearest.second;

            double theta = std::atan2(y_node - y_parent, x_node - x_parent);
            int x_new = static_cast<int>(x_parent + step_size * std::cos(theta));
            int y_new = static_cast<int>(y_parent + step_size * std::sin(theta));

            if (reached) {
                iterations++;
            }

            if (valid_node(canvas, visited, x_new, y_new, distance_threshold)) {
                std::vector<std::pair<int, int>> neighbours;
                for (const auto& p : parent) {
                    if (distance(p.first.first, p.first.second, x_new, y_new) < search_radius) {
                        neighbours.push_back(p.first);
                    }
                }

                double min_cost = cost[{x_parent, y_parent}] + step_size;
                for (const auto& n : neighbours) {
                    double new_cost = cost[n] + distance(n.first, n.second, x_new, y_new);
                    if (new_cost < min_cost) {
                        min_cost = new_cost;
                        x_parent = n.first;
                        y_parent = n.second;
                    }
                }

                if (!valid_edge(canvas, x_parent, y_parent, x_new, y_new)) {
                    continue;
                }

                cost[{x_new, y_new}] = min_cost;

                for (const auto& n : neighbours) {
                    double new_cost = cost[{x_new, y_new}] + distance(n.first, n.second, x_new, y_new);
                    if (new_cost < cost[n]) {
                        if (!valid_edge(canvas, n.first, n.second, x_new, y_new)) {
                            continue;
                        }

                        double previous_cost = cost[n];
                        cost[n] = new_cost;

                        std::pair<int, int> x_parent_neighbour = parent[n];
                        auto& children_list = children[x_parent_neighbour];
                        children_list.erase(std::remove(children_list.begin(), children_list.end(), n), children_list.end());

                        parent[n] = {x_new, y_new};
                        children[{x_new, y_new}].push_back(n);

                        double cost_reduction = previous_cost - new_cost;
                        std::queue<std::pair<int, int>> queue;
                        queue.push(n);
                        while (!queue.empty()) {
                            std::pair<int, int> current = queue.front();
                            queue.pop();
                            if (children.find(current) != children.end()) {
                                for (const auto& child : children[current]) {
                                    cost[child] -= cost_reduction;
                                    queue.push(child);
                                }
                            }
                        }
                    }
                }

                int x_achieved = x_new;
                int y_achieved = y_new;

                int x_adjusted = adjust(x_new, distance_threshold);
                int y_adjusted = adjust(y_new, distance_threshold);

                visited[{x_adjusted, y_adjusted}] = 1;
                parent[{x_new, y_new}] = {x_parent, y_parent};
                children[{x_parent, y_parent}].push_back({x_new, y_new});

                if (!reached) {
                    if (distance(x_new, y_new, x_goal, y_goal) < distance_threshold) {
                        auto end = std::chrono::high_resolution_clock::now();
                        std::chrono::duration<double> elapsed = end - start;
                        std::cout << "Goal reached: " << elapsed.count() << " seconds" << std::endl;

                        children[{x_new, y_new}].push_back({x_goal, y_goal});
                        parent[{x_goal, y_goal}] = {x_new, y_new};

                        int x_final = x_goal;
                        int y_final = y_goal;

                        cost[{x_goal, y_goal}] = cost[{x_new, y_new}] + distance(x_goal, y_goal, x_new, y_new);
                        std::cout << "Current Cost: " << cost[{x_goal, y_goal}] << std::endl;

                        reached = true;
                    }
                }
            }
        }
    }

    int x_final = x_goal;
    int y_final = y_goal;
    std::cout << "Final cost: " << cost[{x_final, y_final}] << std::endl;

    std::vector<std::pair<int, int>> path;
    std::pair<int, int> current = {x_final, y_final};
    while (current != std::make_pair(x_start, y_start)) {
        path.push_back(current);
        current = parent[current];
    }
    path.push_back({x_start, y_start});
    std::reverse(path.begin(), path.end());

    double sum = 0;
    for (size_t i = 0; i < path.size() - 1; ++i) {
        sum += distance(path[i].first, path[i].second, path[i + 1].first, path[i + 1].second);
    }

    std::cout << "Total distance: " << sum << std::endl;
    std::cout << "Difference in cost: " << std::round(cost[{x_final, y_final}] - sum) << std::endl;

    canvas = obstacles();

    cv::circle(canvas, cv::Point(x_start, y_start), 5, cv::Scalar(0, 0, 254), 10);
    cv::circle(canvas, cv::Point(x_goal, y_goal), 5, cv::Scalar(0, 0, 254), 10);

    for (const auto& p : parent) {
        cv::line(canvas, cv::Point(p.first.first, p.first.second), cv::Point(p.second.first, p.second.second), cv::Scalar(0, 255, 0), 5);
        cv::circle(canvas, cv::Point(p.second.first, p.second.second), 5, cv::Scalar(255, 0, 255), 3);
    }

    for (const auto& c : children) {
        for (const auto& child : c.second) {
            cv::circle(canvas, cv::Point(child.first, child.second), 5, cv::Scalar(255, 0, 255), 3);
        }
    }

    for (size_t i = 0; i < path.size() - 1; ++i) {
        cv::line(canvas, cv::Point(path[i].first, path[i].second), cv::Point(path[i + 1].first, path[i + 1].second), cv::Scalar(0, 0, 255), 5);
    }

    int width_resized = width / scale;
    int height_resized = height / scale;
    cv::Mat canvas_resized;
    cv::resize(canvas, canvas_resized, cv::Size(width_resized, height_resized));

    cv::imshow("Canvas", canvas_resized);
    cv::waitKey(0);
    cv::destroyAllWindows();

    return 0;
}

